src/tools/w3dviewer/w3dviewdoc.cpp

```diff
@@ -0,0 +1,957 @@
+                    float vfov = atof(theApp.GetProfileString("Config", "vfov", "0"));
+                    camera->Set_View_Plane(hfov, vfov);
+                }
+
+                if (m_useManualClipPlanes) {
+                    float znear = atof(theApp.GetProfileString("Config", "znear", "0.1F"));
+                    float zfar = atof(theApp.GetProfileString("Config", "zfar", "100.0F"));
+                    camera->Set_Clip_Planes(znear, zfar);
+
+                    if (m_scene != nullptr) {
+                        m_scene->Set_Fog_Range(znear, zfar);
+                        m_scene->Update_Fog_Range();
+                    }
+                }
+            }
+        }
+    }
+}
+
+CDataTreeView *CW3DViewDoc::GetDataTreeView()
+{
+    CMainFrame *frame = static_cast<CMainFrame *>(AfxGetMainWnd());
+
+    if (frame != nullptr) {
+        return static_cast<CDataTreeView *>(frame->m_splitter.GetPane(0, 0));
+    } else {
+        return nullptr;
+    }
+}
+
+CGraphicView *CW3DViewDoc::GetGraphicView()
+{
+    CMainFrame *frame = static_cast<CMainFrame *>(AfxGetMainWnd());
+
+    if (frame != nullptr) {
+        return static_cast<CGraphicView *>(frame->m_splitter.GetPane(0, 1));
+    } else {
+        return nullptr;
+    }
+}
+
+void CW3DViewDoc::UpdateParticleCount()
+{
+    int count = GetNumParticles(nullptr);
+
+    static_cast<CMainFrame *>(AfxGetMainWnd())->UpdateParticleCount(count);
+}
+
+int CW3DViewDoc::GetNumParticles(RenderObjClass *robj)
+{
+    int count = 0;
+
+    if (robj == nullptr) {
+        robj = m_model;
+    }
+
+    if (robj != nullptr) {
+        for (int i = 0; i < robj->Get_Num_Sub_Objects(); i++) {
+            RenderObjClass *o = robj->Get_Sub_Object(i);
+
+            if (o != nullptr) {
+                count += GetNumParticles(o);
+                o->Release_Ref();
+            }
+        }
+
+        if (robj->Class_ID() == RenderObjClass::CLASSID_PARTICLEEMITTER) {
+
+            if (static_cast<ParticleEmitterClass *>(robj)->Peek_Buffer() != nullptr) {
+                count += static_cast<ParticleEmitterClass *>(robj)->Peek_Buffer()->Get_Particle_Count();
+            }
+        }
+    }
+
+    return count;
+}
+
+void CW3DViewDoc::SetRenderObject(RenderObjClass *robj, bool unk1, bool unk2, bool unk3)
+{
+    if (m_scene == nullptr) {
+        return;
+    }
+
+    Ref_Ptr_Release(m_animation);
+
+    if (!unk3 && m_model != nullptr) {
+        RemoveRenderObject(m_model);
+        Ref_Ptr_Release(m_model);
+    }
+
+    m_scene->Remove_All_LOD_Objects();
+
+    if (robj == nullptr) {
+        return;
+    }
+
+    if (unk3) {
+        robj->Set_Animation();
+        Matrix3D tm(true);
+        robj->Set_Transform(tm);
+        SceneClass *scene;
+
+        if (robj->Class_ID() == RenderObjClass::CLASSID_BITMAP2D) {
+            scene = m_textureScene;
+        } else {
+            m_scene->Remove_All_LOD_Objects();

```

Comment:
Why is this called here when it is already called on line 436?

Reply:
I don't know, ask Westwood.